File: Lesson8 Algorithms/Problem2/Solutions/Solution.py
class Test:
    def __init__(self, score, hour):
        self.score = score
        self.hour = hour
    def __lt__(self, other):
        if self.hour == other.hour:
            return other.hour > self.hour
        return self.hour < other.hour

n, m, k = map(int, raw_input().split())
cnt = 0
t = []
for i in range (n):
    x, y = map(int, raw_input().split())
    t.append(Test(x, y))
    cnt += t[i].score


    
ans = 0
t.sort()
for i in range (n):
    if cnt + (m - t[i].score) <= n * k:
        ans += (m - t[i].score) * t[i].hour
        cnt += (m - t[i].score)
    else:
        ans += t[i].hour * (n * k - cnt)
        break

if ans < 0:
    print 0
else:
    print ans

File: Lesson8 Algorithms/Problem2/Solutions/Solution.cpp
#include <iostream>
#include<algorithm>
using namespace std;

struct Test{
    int score, hour;
};
bool cmp(Test a, Test b){
    if (a.hour == b.hour){
        return a.score > b.score;
    }
    return a.hour < b.hour;
}
Test t[100010];
int main()
{
   int n; long m, k, cnt = 0;
   cin>>n>>m>>k;
   for(int i = 0; i < n; i++){
       cin>>t[i].score>>t[i].hour;
       cnt += t[i].score;
   }
   
   if(cnt >= n*k){
       cout<<0<<endl;
       return 0;
   }
   sort(t, t + n, cmp);
   long ans = 0;
   for(int i = 0; i < n; i++){
       if(cnt + (m - t[i].score) <= n*k){
           ans += (m - t[i].score) * t[i].hour;
           cnt += (m - t[i].score);
       }
       else{
           ans += (n*k - cnt) * t[i].hour;
           break;
       }
   }
   cout<<ans<<endl;
   return 0;
}
